### PR TITLE
parser: support `ALTER PARTITION` clause

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

As title, it follows the first task of https://github.com/pingcap/tidb/issues/18405.

### Check List

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: parser.y [hunk 1]
```diff
@@ -1339,6 +1349,63 @@ AlterTableStmt:
...
+	{
+		$$ = ast.PlacementRoleLearner
+	}
+|	"VOTER"
+	{
+		$$ = ast.PlacementRoleVoter
+	}
+
+PlacementOpts:
+	"LABEL" "=" stringLit "ROLE" "=" PlacementRole "COUNT" "=" LengthNum
```
Comment (djshow832) [inline parser.y:1371]: Users may use another order, such as `ROLE=... COUNT=... LABEL=...`

File: ast/ddl.go [hunk 2]
```diff
@@ -2530,6 +2533,20 @@ func (n *AlterTableSpec) Restore(ctx *format.RestoreCtx) error {
...
+			return errors.Errorf("Maybe partition options are combined.")
+		}
+
+		ctx.WriteKeyWord("ALTER PARTITION ")
+		ctx.WriteName(n.PartitionNames[0].O)
+		ctx.WritePlain(" ")
+
+		for i, spec := range n.PlacementSpecs {
+			if err := spec.Restore(ctx); err != nil {
+				return errors.Annotatef(err, "An error occurred while restore AlterTableSpec.PLacementSpecs[%d]", i)
```
Comment (djshow832) [inline ast/ddl.go:2547]:
```suggestion
				return errors.Annotatef(err, "An error occurred while restore AlterTableSpec.PlacementSpecs[%d]", i)
```

File: ast/ddl.go [hunk 3]
```diff
@@ -2720,6 +2737,12 @@ func (n *AlterTableSpec) Restore(ctx *format.RestoreCtx) error {
 		case IndexVisibilityInvisible:
 			ctx.WriteKeyWord(" INVISIBLE")
 		}
+	case AlterTablePlacement:
+		for i, spec := range n.PlacementSpecs {
+			if err := spec.Restore(ctx); err != nil {
+				return errors.Annotatef(err, "An error occurred while restore AlterTableSpec.PLacementSpecs[%d]", i)
```
Comment (djshow832) [inline ast/ddl.go:2743]:
```suggestion
				return errors.Annotatef(err, "An error occurred while restore AlterTableSpec.PlacementSpecs[%d]", i)
```

File: parser.y [hunk 4]
```diff
@@ -1479,6 +1546,14 @@ AlterTableSpec:
 			Num:             getUint64FromNUM($6),
 		}
 	}
+|	"ALTER" "PARTITION" Identifier PlacementSpecList %prec lowerThanComma
+	{
+		$$ = &ast.AlterTableSpec{
+			Tp:             ast.AlterTableAlterPartition,
+			PartitionNames: []model.CIStr{model.NewCIStr($3)},
+			PlacementSpecs: $4.([]*ast.PlacementSpec),
```
Comment (djshow832) [inline parser.y:1554]: If placement rules on partitions are also stored in `PlacementSpecs`, how can you figure out whether they belong to the table or the partition?
E.g.
```sql
ALTER TABLE t
        ALTER PLACEMENT ..., 
        ALTER PARTITION ALTER PLACEMENT ...
```

Comment (xhebox) [reply parser.y:1554]: `Tp = AlterTableAlterPartition` for `ALTER PARTITION ALTER PLACEMENT...`, and `Tp = AlterTablePlacement` for `ALTER PLACEMENT...`. And they should belong to two different `AlterTableSpec`. I think that is enough.

File: ast/ddl.go [hunk 5]
```diff
@@ -3426,3 +3449,68 @@ func (n *FlashBackTableStmt) Accept(v Visitor) (Node, bool) {
...
+func (n *PlacementSpec) Restore(ctx *format.RestoreCtx) error {
+	switch n.Tp {
+	case PlacementAdd:
+		ctx.WriteKeyWord("ADD PLACEMENT ")
+	default:
+		return errors.Errorf("invalid PlacementActionType: %d", n.Tp)
+	}
+
+	ctx.WriteKeyWord("LABEL")
+	ctx.WritePlainf("=\"%s\" ", n.Labels)
```
Comment (kennytm) [inline ast/ddl.go:3486]:
```suggestion
	ctx.WritePlain("=")
	ctx.WriteString(n.Labels)
```

you don't want to restore to an invalid string if the label contains a `'` or `"`.

Comment (xhebox) [reply ast/ddl.go:3486]: Thanks.

File: ast/ddl.go [hunk 6]
```diff
@@ -3426,3 +3449,70 @@ func (n *FlashBackTableStmt) Accept(v Visitor) (Node, bool) {
...
+	default:
+		return errors.Errorf("invalid PlacementActionType: %d", n.Tp)
+	}
+
+	ctx.WriteKeyWord("LABEL")
+	ctx.WritePlain("=")
+	ctx.WriteString(n.Labels)
+	ctx.WritePlain(" ")
+
+	ctx.WriteKeyWord("ROLE")
```
Comment (kennytm) [inline ast/ddl.go:3490]:
```suggestion

	ctx.WriteKeyWord(" ROLE")
```

File: ast/ddl.go [hunk 7]
```diff
@@ -3426,3 +3449,70 @@ func (n *FlashBackTableStmt) Accept(v Visitor) (Node, bool) {
...
+	case PlacementRoleLearner:
+		ctx.WriteKeyWord("LEARNER")
+	case PlacementRoleVoter:
+		ctx.WriteKeyWord("VOTER")
+	default:
+		return errors.Errorf("invalid PlacementRole: %d", n.Role)
+	}
+	ctx.WritePlain(" ")
+
+	ctx.WriteKeyWord("COUNT")
```
Comment (kennytm) [inline ast/ddl.go:3506]:
```suggestion
	ctx.WriteKeyWord(" COUNT")
```

File: parser.y [hunk 8]
```diff
@@ -642,6 +644,10 @@ import (
 	optRuleBlacklist      "OPT_RULE_BLACKLIST"
 	jsonObjectAgg         "JSON_OBJECTAGG"
 	tls                   "TLS"
+	follower              "FOLLOWER"
+	leader                "LEADER"
+	learner               "LEARNER"
```
Comment (kennytm) [inline parser.y:649]: (not review — gotta say these two terms looks too similar 🙃)

File: ast/ddl.go [hunk 9]
```diff
@@ -3426,3 +3449,68 @@ func (n *FlashBackTableStmt) Accept(v Visitor) (Node, bool) {
...
+	PlacementAdd PlacementActionType = iota + 1
+)
+
+type PlacementRole int
+
+const (
+	PlacementRoleFollower PlacementRole = iota + 1
+	PlacementRoleLeader
+	PlacementRoleLearner
+	PlacementRoleVoter
```
Comment (djshow832) [inline ast/ddl.go:3465]: May this order would be more natural:
```suggestion
	PlacementRoleLeader PlacementRole = iota + 1
	PlacementRoleFollower
	PlacementRoleLearner
	PlacementRoleVoter
```

File: parser_test.go [hunk 10]
```diff
@@ -2570,6 +2571,18 @@ func (s *testParserSuite) TestDDL(c *C) {
...
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT LABEL='+zone=zone1,-zone=zone2' ROLE=LEADER COUNT=1", true, "ALTER TABLE `t` ALTER PARTITION `p` ADD PLACEMENT LABEL='+zone=zone1,-zone=zone2' ROLE=LEADER COUNT=1"},
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT LABEL='+zone=zone1' ROLE=FOLLOWER COUNT=1", true, "ALTER TABLE `t` ALTER PARTITION `p` ADD PLACEMENT LABEL='+zone=zone1' ROLE=FOLLOWER COUNT=1"},
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT ROLE=LEARNER LABEL='+zone=zone1' COUNT=1", true, "ALTER TABLE `t` ALTER PARTITION `p` ADD PLACEMENT LABEL='+zone=zone1' ROLE=LEARNER COUNT=1"},
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT COUNT=1 LABEL='+zone=zone1' ROLE=VOTER", true, "ALTER TABLE `t` ALTER PARTITION `p` ADD PLACEMENT LABEL='+zone=zone1' ROLE=VOTER COUNT=1"},
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT LABEL= ROLE=follower COUNT=1;", false, ""},
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT LABEL='+zone=gh' ROLE=follower COUNT=-1;", false, ""},
+		{"ALTER TABLE t ALTER PARTITION p ADD PLACEMENT LABEL='+zone=gh' ROLE=follower COUNT=0;", false, ""},
+		{"ALTER TABLE t ALTER PARTITION p", false, ""},
+
```
Comment (djshow832) [inline parser_test.go:2585]: Add a test case that has duplicate LABEL, ROLE or COUNT.

Comment (xhebox) [reply parser_test.go:2589]: done

**Comment by djshow832** [comment]
Please give another LGTM @kennytm
